### PR TITLE
zypper_repository: fix idempotency on adding repo with releasever and basearch variables

### PR DESCRIPTION
**Pull request by aminvakil**
##### SUMMARY
Fixes https://github.com/ansible-collections/community.general/issues/1985

##### ISSUE TYPE
- Bugfix Pull Request

##### COMPONENT NAME
zypper_repository

##### ADDITIONAL INFORMATION
I'm not sure about this being a bugfix or feature request, please let me know if I should change it.

### CODE REVIEW

**Comment by aminvakil** [comment]
`parse_repos` function which checks current defined repos converts `$releasever` to opensuse version right now, therefore it cannot remove the added repository by `$releasever` right now.

Output of `parse_repos`:
```
fatal: [testhost]: FAILED! => {
    "changed": false,
    "invocation": {
        "module_args": {
            "auto_import_keys": false,
            "autorefresh": true,
            "description": null,
            "disable_gpg_check": false,
            "enabled": true,
            "name": null,
            "overwrite_multiple": false,
            "priority": null,
            "repo": "http://download.opensuse.org/repositories/devel:/languages:/ruby/openSUSE_Leap_$releasever/",
            "runrefresh": false,
            "state": "absent"
        }
    },
    "msg": "<?xml version='1.0'?>\n<stream>\n<repo-list>\n<repo alias=\"releaseverrepo\" name=\"releaseverrepo\" priority=\"99\" enabled=\"1\" autorefresh=\"1\" gpgcheck=\"1\" repo_gpgcheck=\"1\" pkg_gpgcheck=\"1\">\n<url>http://download.opensuse.org/repositories/devel:/languages:/ruby/openSUSE_Leap_15.2/</url>\n</repo>\n<repo alias=\"repo-debug\" name=\"Debug Repository\" priority=\"99\" enabled=\"0\" autorefresh=\"1\" gpgcheck=\"1\" repo_gpgcheck=\"1\" pkg_gpgcheck=\"1\">\n<url>http://download.opensuse.org/debug/distribution/leap/15.2/repo/oss/</url>\n</repo>\n<repo alias=\"repo-debug-non-oss\" name=\"Debug Repository (Non-OSS)\" priority=\"99\" enabled=\"0\" autorefresh=\"1\" gpgcheck=\"1\" repo_gpgcheck=\"1\" pkg_gpgcheck=\"1\">\n<url>http://download.opensuse.org/debug/distribution/leap/15.2/repo/non-oss/</url>\n</repo>\n<repo alias=\"repo-debug-update\" name=\"Update Repository (Debug)\" priority=\"99\" enabled=\"0\" autorefresh=\"1\" gpgcheck=\"1\" repo_gpgcheck=\"1\" pkg_gpgcheck=\"1\">\n<url>http://download.opensuse.org/debug/update/leap/15.2/oss/</url>\n</repo>\n<repo alias=\"repo-debug-update-non-oss\" name=\"Update Repository (Debug, Non-OSS)\" priority=\"99\" enabled=\"0\" autorefresh=\"1\" gpgcheck=\"1\" repo_gpgcheck=\"1\" pkg_gpgcheck=\"1\">\n<url>http://download.opensuse.org/debug/update/leap/15.2/non-oss/</url>\n</repo>\n<repo alias=\"repo-non-oss\" name=\"Non-OSS Repository\" priority=\"99\" enabled=\"1\" autorefresh=\"1\" gpgcheck=\"1\" repo_gpgcheck=\"1\" pkg_gpgcheck=\"1\">\n<url>http://download.opensuse.org/distribution/leap/15.2/repo/non-oss/</url>\n</repo>\n<repo alias=\"repo-oss\" name=\"Main Repository\" priority=\"99\" enabled=\"1\" autorefresh=\"1\" gpgcheck=\"1\" repo_gpgcheck=\"1\" pkg_gpgcheck=\"1\">\n<url>http://download.opensuse.org/distribution/leap/15.2/repo/oss/</url>\n</repo>\n<repo alias=\"repo-source\" name=\"Source Repository\" priority=\"99\" enabled=\"0\" autorefresh=\"1\" gpgcheck=\"1\" repo_gpgcheck=\"1\" pkg_gpgcheck=\"1\">\n<url>http://download.opensuse.org/source/distribution/leap/15.2/repo/oss/</url>\n</repo>\n<repo alias=\"repo-source-non-oss\" name=\"Source Repository (Non-OSS)\" priority=\"99\" enabled=\"0\" autorefresh=\"1\" gpgcheck=\"1\" repo_gpgcheck=\"1\" pkg_gpgcheck=\"1\">\n<url>http://download.opensuse.org/source/distribution/leap/15.2/repo/non-oss/</url>\n</repo>\n<repo alias=\"repo-update\" name=\"Main Update Repository\" priority=\"99\" enabled=\"1\" autorefresh=\"1\" gpgcheck=\"1\" repo_gpgcheck=\"1\" pkg_gpgcheck=\"1\">\n<url>http://download.opensuse.org/update/leap/15.2/oss/</url>\n</repo>\n<repo alias=\"repo-update-non-oss\" name=\"Update Repository (Non-Oss)\" priority=\"99\" enabled=\"1\" autorefresh=\"1\" gpgcheck=\"1\" repo_gpgcheck=\"1\" pkg_gpgcheck=\"1\">\n<url>http://download.opensuse.org/update/leap/15.2/non-oss/</url>\n</repo>\n</repo-list>\n</stream>\n"
```

It can be seen that it's converting all `$releasever` to 15.2 and it's not related to this PR.

So this PR only fixes the idempotency.

**Comment by aminvakil** [comment]
@dev-zero Could you please review this PR and see if this fixes your issue or not?

**Comment by felixfontein** [comment]
You need a changelog fragment :)

File: changelogs/fragments/2722-zypper_repository-fix_idempotency_on_adding_repo_with_releasever.yml [hunk 1]
```diff
@@ -0,0 +1,4 @@
+---
+bugfixes:
+  - zypper_repository - fix idempotency on adding repo with releasever
```
Comment (felixfontein) [inline changelogs/fragments/2722-zypper_repository-fix_idempotency_on_adding_repo_with_releasever.yml:3]:
```suggestion
  - zypper_repository - fix idempotency on adding repository with ``$releasever``
```

File: plugins/modules/packaging/os/zypper_repository.py [hunk 2]
```diff
@@ -186,6 +186,10 @@ def _repo_changes(realrepo, repocmp):
             valold = str(repocmp[k] or "")
             valnew = v or ""
             if k == "url":
+                cmd = ('rpm -q --qf "%{version}" -f /etc/os-release')
```
Comment (felixfontein) [inline plugins/modules/packaging/os/zypper_repository.py:189]:
```suggestion
                cmd = ['rpm', '-q', '--qf', '%{version}', '-f', '/etc/os-release']
```

File: plugins/modules/packaging/os/zypper_repository.py [hunk 3]
```diff
@@ -186,6 +186,10 @@ def _repo_changes(realrepo, repocmp):
             valold = str(repocmp[k] or "")
             valnew = v or ""
             if k == "url":
+                cmd = ['rpm', '-q', '--qf', '%{version}', '-f', '/etc/os-release']
+                rc, stdout, stderr = module.run_command(cmd, check_rc=True)
```
Comment (aminvakil) [inline plugins/modules/packaging/os/zypper_repository.py:190]: I think the command is pretty safe, but should I add a check that if `rc != 0` skip the whole replace?

Comment (felixfontein) [reply plugins/modules/packaging/os/zypper_repository.py:190]: `check_rc=True` will already make the module fail, so `rc` will always be `== 0` here.

Comment (aminvakil) [reply plugins/modules/packaging/os/zypper_repository.py:190]: Yes, I meant skip instead of fail, so no one notices this change :)

Comment (felixfontein) [reply plugins/modules/packaging/os/zypper_repository.py:190]: Hmm, I think it's generally not a good idea to ignore errors, so it's probably better to not fail. But I have no idea how safe that command is :)

Comment (aminvakil) [reply plugins/modules/packaging/os/zypper_repository.py:190]: Fine, I'll let it be as it is.
It's pretty safe IMO, I can also change it to use `grep` to retrieve version directly from `/etc/os-release`, but I still think this is better.

**Comment by dev-zero** [comment]
I haven't tested the patch so far, sorry for that, but until now I encountered the following vars in repo urls:

* `$releasever` ... recommended by openSUSE to be used to make upgrades easier
* `$basearch`  ... used by packagecloud, for example here: https://packagecloud.io/netdata/netdata/install#manual-zypper

other vars which according to https://doc.opensuse.org/projects/libzypp/HEAD/zypp-repovars.html are also allowed (builtin) but I haven't encountered/used so far are:

* `$arch`
* `$releasever_minor`, `$releasever_major`

**Comment by aminvakil** [comment]
>`$basearch`  ... used by packagecloud, for example here: https://packagecloud.io/netdata/netdata/install#manual-zypper
> 
Added in https://github.com/ansible-collections/community.general/pull/2722/commits/042ea9f729b7e870eef68b485059e3b200441d61.
> 
> other vars which according to https://doc.opensuse.org/projects/libzypp/HEAD/zypp-repovars.html are also allowed (builtin) but I haven't encountered/used so far are:
> 
>     * `$arch`
> 
>     * `$releasever_minor`, `$releasever_major`

I have not used nor seen in documentations and tutorials to advice on using them, let's not add them for now.

File: plugins/modules/packaging/os/zypper_repository.py [hunk 4]
```diff
@@ -186,6 +186,14 @@ def _repo_changes(realrepo, repocmp):
...
+                valnew = valnew.replace('$releasever', stdout)
+                valold = valold.replace('$releasever', stdout)
+                cmd = ['rpm', '-q', '--qf', '%{arch}', '-f', '/etc/os-release']
+                rc, stdout, stderr = module.run_command(cmd, check_rc=True)
+                valnew = valnew.replace('$basearch', stdout)
+                valold = valold.replace('$basearch', stdout)
```
Comment (felixfontein) [inline plugins/modules/packaging/os/zypper_repository.py:196]: How about only running these commands `if '$releasever' in valold or '$releasever' in vernew:` for the first, and similar for the second? Then people who don't use these variables do not get extra runtime for executing `rpm` for each of them.

Comment (aminvakil) [reply plugins/modules/packaging/os/zypper_repository.py]: Nice idea. Added in https://github.com/ansible-collections/community.general/pull/2722/commits/b1db85c5043ecd9332d4b5c112d9cf6a9e67106f.

**Review by felixfontein — APPROVED** [review]

**Comment by felixfontein** [comment]
@aminvakil thanks for implementing this!
@dev-zero thanks for reviewing!

**Pull request merged**